### PR TITLE
27 daskwatershed

### PR DESCRIPTION
**Pull request by rafa-guedes**
Make watershed partitioning dask-able. Partitioning was implemented with apply_ufuncs and now fully support dask arrays as inputs and can handle very large spectra that would otherwise not fit in memory, as long as they are chunked appropriately.

### CODE REVIEW

File: wavespectra/core/watershed.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        - dset (xr.DataArray, xr.Dataset): Spectra array or dataset in wavespectra convention.
+        - wspd (xr.DataArray, str): Wind speed DataArray or variable name in dset.
+        - wdir (xr.DataArray, str): Wind direction DataArray or variable name in dset.
+        - dpt (xr.DataArray, str): Depth DataArray or the variable name in dset.
+        - swells (int): Number of swell partitions to compute.
+        - agefac (float): Age factor.
+        - wscut (float): Wind speed cutoff.
+
+    Returns:
+        - dspart (xr.Dataset): Partitioned spectra dataset with extra dimension.
```
Comment (tomdurrant) [inline wavespectra/core/watershed.py:100]: It looks like the partitions are populated with zero partitions if <swells number of swells are present. Might be worth putting in a note to that effect give the change from max_swells

File: wavespectra/specarray.py [hunk 2]
```diff
@@ -701,11 +658,9 @@ def partition(
         wsp_darr,
         wdir_darr,
         dep_darr,
+        swells=3,
```
Comment (tomdurrant) [inline wavespectra/specarray.py:661]: The default here of 3 looks sensible, and makes sense to reduce given the change from max_swells - just making sure it been a deliberate change.

**Review by tomdurrant — APPROVED** [review]
Couple of very minor comments, but all looks sensible to me. All tests pass on my system

File: wavespectra/core/watershed.py (hunk 1 above)
Comment (rafa-guedes) [reply wavespectra/core/watershed.py:100]: Yep we should made this change more clear. Do you mean put a note in the release notes or in the docstring?

The actual behaviour though has not really changed with the new argument. With the previous version, `max_swells` controlled the maximum number for the full dataset and specific spectrum in the dataset that did not have enough partitions were also filled with zeros. I thought of changing the argument name / description though as we now vectorise the calculations and require a fixed number of partitions so output can be constructed. And `max_swells` is a bit misleading given some spectrum will not have those numbers.

File: wavespectra/specarray.py (hunk 2 above)
Comment (rafa-guedes) [reply wavespectra/specarray.py:661]: Yes it was deliberate change, we more commonly request up to 3 swells rather than 5.

**Pull request merged**